[Toolkit][Shadcn] Add DropdownMenu component

daFish opened the pull request:
| Q              | A
| -------------- | ---
| Bug fix?       | no
| New feature?   | yes <!-- please update src/**/CHANGELOG.md files -->
| Deprecations?  | no <!-- if yes, also update UPGRADE-*.md and src/**/CHANGELOG.md -->
| Documentation? | yes <!-- required for new features, or documentation updates -->
| Issues         | Tracked in #3233 
| License        | MIT

Add DropdownMenu as a component with support for checkboxes, radio group, and dialog. It utilizes the already available components like Kbd and Dialog. The Button Group component is also updated to use the new component.

ref: https://ui.shadcn.com/docs/components/dropdown-menu

daFish committed:
feat(toolkit): allow custom classes to be passed from the outside

daFish committed:
feat(toolkit): add dropdown-menu component for shadcn kit

daFish committed:
feat(toolkit): add support for dropdown-menu in button-group

daFish commented:
I have also added support for dropdown-menu in button group.

/cc @bernard-ng

Kocal commented:
Hi!

I started to look at your work, and thanks you so much for that!

However, I find there is too much logic in the `dropdown_menu_controller.js`, and I feel like it could be super-simplified by using these new features:
- https://developer.chrome.com/blog/command-and-commandfor
- https://developer.mozilla.org/en-US/docs/Web/CSS/Guides/Anchor_positioning?

Thanks!

Kocal commented on src/Toolkit/kits/shadcn/dropdown-menu/manifest.json:
```diff
@@ -0,0 +1,15 @@
+    "dependencies": {
+        "recipe": ["kbd"],
+        "composer": ["twig/extra-bundle", "twig/html-extra:^3.12.0", "tales-from-a-dev/twig-tailwind-extra:^1.0.0"],
+        "importmap": ["@floating-ui/dom"]
```
Where it is imported?

daFish replied:
Will be removed.

Kocal commented on src/Toolkit/kits/shadcn/dropdown-menu/manifest.json:
```diff
@@ -0,0 +1,15 @@
+        "templates/": "templates/"
+    },
+    "dependencies": {
+        "recipe": ["kbd"],
```
Let's remove this line, as kbd is not required to render the dropdown menu

daFish replied:
Done.

Kocal commented on src/Toolkit/kits/shadcn/dropdown-menu/templates/components/DropdownMenu/Content.html.twig:
```diff
@@ -0,0 +1,24 @@
+{# @block content The default block #}
+{%- props side = 'bottom', align = 'start' -%}
+{%- set style = html_cva(
+    base: 'absolute z-50 min-w-[8rem] overflow-hidden rounded-md border bg-popover p-1 text-popover-foreground shadow-md outline-none',
```
IMHO we should re-use the opening transitions from the AlertDialog/Dialog components

daFish replied:
Will be refactored.

daFish commented:
@Kocal How to handle custom styles? Should I import them in the controller or is this something to be handled in user land?

daFish commented on ux.symfony.com/importmap.php:
```diff
@@ -218,4 +221,16 @@
     'react-dom/client' => [
         'version' => '18.3.1',
     ],
+    '@floating-ui/dom' => [
```
All `@floating-ui`-dependencies must be removed.

Kocal commented:
Hey!

Sorry for the long delay, but I've been pretty busy with https://github.com/symfony/ux/pull/3299, which added a tons of changes for documentation and code examples. 😅 

They are now shared between this repo and https://github.com/symfony/ux.symfony.com

I will review your PR when possible, can you apply the changes in the meantime?

Thanks! 🙏🏻

daFish commented:
@Kocal No worries, Hugo. I just rebased but missed out on the changes you pushed. 😬

daFish committed:
test(toolkit): update button-group snapshots

daFish committed:
test(toolkit): update input-group snapshot

Kocal commented:
Looks like something wrong happened when rebased 😱 

Don't worry, I will take it from here, thanks for your contribution :)

daFish commented:
@Kocal Thanks. I think I messed up the rebase because I didn't expected new commits in this PR and I force pushed from my side.

Kocal commented:
(I've been pretty busy lately, but I'll try to take a look at it next week :))

rvmourik commented:
> (I've been pretty busy lately, but I'll try to take a look at it next week :))

Hi @Kocal, seeing a lot of positive changes to symfony UX lately, thank you for that. Is there any progress on this specific one, I am really eager to use it for the button groups. Kind of the last thing holding me back from switching our SaaS-application to using the Toolkit. The other components are already there 👍 

Thanks for your great work

Kocal commented:
Hi, Symfony UX 3.0 has been released, and the `2.x` branch is no longer maintained.

Could you please retarget this PR to the `3.x` branch instead? That way we can review and merge it in the currently supported version.

Thanks!

Status: Needs Work

daFish commented:
@Kocal Rebased against `3.x` and also changed target branch.

daFish commented:
Status: Needs Work

daFish commented:
@Kocal I need your opinion regarding this PR and #3469. I wouldn't mind closing this one in favor of the other. The goal remains the same.

Kocal committed:
[Toolkit][Shadcn] Fix dialog opening on initial render when open is false

The root template rendered data-dialog-open-value="{{ open }}", and
Twig turns a false boolean into an empty string, producing
data-dialog-open-value="". Stimulus treats an empty string as boolean
true, so a dialog left at its default open = false called showModal
on connect. Render an explicit 'true'/'false' string instead.

Kocal committed:
[Toolkit][Shadcn] Add dropdown-menu recipe

Add the shadcn dropdown-menu recipe: DropdownMenu plus Trigger,
Content, Item, CheckboxItem, RadioGroup, RadioItem, Group, Label,
Separator, Shortcut, Sub, SubTrigger and SubContent, covering menus,
checkbox items, radio groups, labels, groups, separators, keyboard
shortcuts and nested submenus.

A Stimulus controller wires the interaction: the trigger opens the
menu, outside clicks or Escape close it (Escape returns focus to the
trigger), opening moves focus to the first item, and arrow keys, Home
and End navigate items with wrapping and disabled-item skipping.
Submenus open on hover or focus. The menu positions itself on a
chosen side/alignment of the trigger with a fade/scale transition and
supports RTL.

Follows current Toolkit conventions: README-only docs with
live-preview examples, literal state/ARIA attributes, classes merged
via tailwind_classes and html_cva, and parent/descendant wiring via
provide()/inject().

Co-authored-by: Marcus Stöhr <ms@pond-solutions.de>
